controller/object_store_controller.go

```diff
@@ -0,0 +1,1159 @@
+	store *longhorn.ObjectStore,
+	pvc *corev1.PersistentVolumeClaim,
+) (*longhorn.Volume, *longhorn.ObjectStore, error) {
+	vol, err := osc.ds.GetVolume(genPVName(store))
+	if err == nil {
+		return vol, store, nil
+	} else if datastore.ErrorIsNotFound(err) {
+		vol, err = osc.createVolume(store, pvc)
+		if err != nil {
+			return nil, store, errors.Wrap(err, "failed to create longhorn volume")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store.Status.State = longhorn.ObjectStoreStateStarting
+		}
+		return vol, store, nil
+	}
+	return nil, store, err
+}
+
+func (osc *ObjectStoreController) checkVolume(vol *longhorn.Volume) error {
+	if vol.Status.Robustness == longhorn.VolumeRobustnessFaulted {
+		return errors.New(fmt.Sprintf("volume %v has failed", vol.Name))
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) getOrCreatePV(
+	store *longhorn.ObjectStore,
+	volume *longhorn.Volume,
+) (*corev1.PersistentVolume, *longhorn.ObjectStore, error) {
+	pv, err := osc.ds.GetPersistentVolume(genPVName(store))
+	if err == nil {
+		return pv, store, nil
+	} else if datastore.ErrorIsNotFound(err) {
+		pv, err = osc.createPV(store, volume)
+		if err != nil {
+			return nil, store, errors.Wrap(err, "failed to create persistent volume")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store.Status.State = longhorn.ObjectStoreStateStarting
+		}
+		return pv, store, nil
+	}
+	return nil, store, err
+}
+
+func (osc *ObjectStoreController) checkPV(pv *corev1.PersistentVolume) error {
+	if pv.Status.Phase != corev1.VolumeBound {
+		return errors.New(fmt.Sprintf("PV %v not bound", pv.Name))
+	}
+	return nil
+}
+
+func (osc *ObjectStoreController) getOrCreateDeployment(store *longhorn.ObjectStore) (*appsv1.Deployment, *longhorn.ObjectStore, error) {
+	dpl, err := osc.ds.GetDeployment(store.Name)
+	if err == nil {
+		return dpl, store, nil
+	} else if datastore.ErrorIsNotFound(err) {
+		dpl, err = osc.createDeployment(store)
+		if err != nil {
+			return nil, store, errors.Wrap(err, "failed to create deployment")
+		} else if store.Status.State != longhorn.ObjectStoreStateStarting {
+			store.Status.State = longhorn.ObjectStoreStateStarting
+		}
+		return dpl, store, nil
+	}
+	return nil, store, err
+}
+
+func (osc *ObjectStoreController) checkDeployment(deployment *appsv1.Deployment, store *longhorn.ObjectStore) error {
+	if *deployment.Spec.Replicas != 1 {
+		deployment.Spec.Replicas = int32Ptr(1)
+		osc.ds.UpdateDeployment(deployment)
+		return errors.New("deployment just scaled")
+	} else if deployment.Status.Replicas == 0 || deployment.Status.UnavailableReplicas > 0 {
+		return errors.New("deployment not ready")
+	}
+
+	if store.Spec.Image != "" && deployment.Spec.Template.Spec.Containers[0].Image != store.Spec.Image {
+		deployment.Spec.Template.Spec.Containers[0].Image = store.Spec.Image
+		osc.ds.UpdateDeployment(deployment)
+		store.Status.State = longhorn.ObjectStoreStateStarting
+	}
```

Comment:
By default will this deployment use a rolling upgrade so that there is always a running and compatible pod? If No, will this lead to a temporary service down?

Reply:
I think currently s3gw cannot have 2 active pods accessing the volume at the same time. The code flow here is inside `handleStarting()` implies that we only change the image the next time the ObjectStore starts. It means that the ObjectStore is not currently running.

Reply:
The deployment must always have either 0 or 1 replica. There must never be multiple replicas.
There is no rolling upgrade as the old instance needs to be shut down before a new one can be started. It will lead to a temporary service down.

Reply:
> The code flow here is inside handleStarting() implies that we only change the image the next time the ObjectStore starts. It means that the ObjectStore is not currently running.

No. The `checkDeployment` function is called both inside the `handleRunning` and `handleStarting` state handlers. Thus the image can be changed on a running object store too. But since the new image can only be rolled out by stopping the old instance first and then starting the new instance, the deployment will have unavailable replicas. In order to avoid mistaking this for an error, we transition to `starting` state while this takes place.

Reply:
> There is no rolling upgrade as the old instance needs to be shut down before a new one can be started. It will lead to a temporary service down.

Is this temporary service down perceptible for users or callers? In other words, will it fail the in-fly IOs?

Reply:
S3 is layered on top of HTTP and as such is a cacheable protocol. In addition to that operations can be retried and their timeout is fairly generous by default, so the actual impact on users is debatable.
@l-mb and @jecluis can explain in much greater detail why there can only be one replica and how the S3 protocol can be leveraged to minimize user impact.
I can tell you right away however that the restriction to one replica is not going to be lifted before LH 1.6 releases, if ever.